dsp/modules/premai.py

```diff
@@ -0,0 +1,151 @@
+    if "rate limits" in details.message:
+        return False
+    return True
+
+
+class PremAI(LM):
+    """Wrapper around Prem AI's API."""
+
+    def __init__(
+        self,
+        model: str,
+        project_id: int,
+        api_key: str,
+        base_url: Optional[str] = None,
+        session_id: Optional[int] = None,
+        **kwargs,
+    ) -> None:
+        """Parameters
+
+        model: str
+            The name of model name
+        project_id: int
+            "The project ID in which the experiments or deployments are carried out. can find all your projects here: https://app.premai.io/projects/"
+        api_key: str
+            Prem AI API key, to connect with the API.
+        session_id: int
+            The ID of the session to use. It helps to track the chat history.
+        **kwargs: dict
+            Additional arguments to pass to the API provider
+        """
+        super().__init__(model)
+        self.kwargs = kwargs if kwargs == {} else self.kwargs
+
+        self.project_id = project_id
+        self.session_id = session_id
+
+        if base_url is not None:
+            self.client = premai.Prem(api_key=api_key, base_url=base_url)
+        else:
+            self.client = premai.Prem(api_key=api_key)
+        self.provider = "premai"
+
+        self.kwargs = {
+            "model": model,
+            "temperature": 0.17,
+            "max_tokens": 150,
+            **kwargs,
+        }
+        if session_id is not None:
+            kwargs["sesion_id"] = session_id
+
+    def _get_all_kwargs(self, **kwargs) -> dict:
+        other_kwargs = {
+            "seed": None,
+            "logit_bias": None,
+            "tools": None,
+            "system_prompt": None,
+        }
+        all_kwargs = {
+            **self.kwargs,
+            **other_kwargs,
+            **kwargs,
+        }
+
+        _keys_that_cannot_be_none = [
+            "system_prompt",
+            "frequency_penalty",
+            "presence_penalty",
+            "tools",
+            "model",
+        ]
+
+        for key in _keys_that_cannot_be_none:
+            if all_kwargs.get(key) is None:
+                all_kwargs.pop(key, None)
+        return all_kwargs
+
+    def basic_request(self, prompt, **kwargs) -> str:
+        """Handles retrieval of completions from Prem AI whilst handling API errors."""
+        all_kwargs = self._get_all_kwargs(**kwargs)
+        message = []
+
+        if "system_prompt" in all_kwargs:
+            message.append({"role": "system", "content": all_kwargs["system_prompt"]})
+        message.append({"role": "user", "content": prompt})
+
+        response = self.client.chat.completions.create(
+            project_id=self.project_id,
+            messages=message,
+            **all_kwargs,
+        )
+        if not response.choices:
+            raise ChatPremAPIError("ChatResponse must have at least one candidate")
+
+        return response.choices[0].message.content or ""
```

Comment:
would it be better to raise this error if the content field is not filled?

Reply:
I mean, yes we can, but we took an assumption that, if `response.choices` exists, then this will also mean that, there should be some response that should exist. If not then there could be any internal error from Prem's side. So, for that we handled it in the above way where if this thing does not exist, then just return ""


what do you think?

Reply:
My comment as more on if `response.choices[0].message.content` is empty, this seems like it would be some incomplete or unexpected response right? and that would be better to raise as a flag over returning `""`.

Reply:
I see, that makes sense. Doing those changes